El trabajo esta basado en diferentes competencias y retos sobre sistemas de recomendaciones.
sitio: https://www.datasource.ai/
A partir del trabajo de https://github.com/PacktPublishing/Hands-On-Data-Science-for-Marketing
Libros: 

## Reto 
### Construyendo Una Sistema de Recomendación de Productos Con Filtrado Colaborativo
### Construyendo Una Sistema de Recomendación de Productos Con Filtrado en el contenido
Vamos a profundizar en la construcción de un sistema de recomendación de productos con el que podemos dirigirnos mejor a los clientes, utilizando recomendaciones de productos que se adaptan a los clientes individuales. Los estudios han demostrado que las recomendaciones de productos personalizadas mejoran las tasas de conversión y de retención de clientes.

Un sistema de recomendación de productos es un sistema cuyo objetivo es predecir y recopilar una lista de artículos que probablemente comprará un cliente. Los sistemas de recomendación han ganado mucha popularidad en los últimos años y se han desarrollado e implementado para diversos casos de uso comercial

Por ejemplo,

El proveedor de servicios multimedia Netflix utiliza sistemas de recomendación para recomendar a los usuarios individuales películas o programas de televisión que es probable que vean.
La empresa de comercio electrónico Amazon utiliza sistemas de recomendación para predecir y mostrar una lista de productos que el cliente probablemente comprará
El servicio de streaming de música, Pandora, utiliza sistemas de recomendación de música para sus oyentes.
El uso de un sistema de recomendación no se detiene aquí. También puede utilizarse para recomendar a los usuarios artículos, noticias o libros relacionados
Con el potencial de ser utilizados en una variedad de áreas, los sistemas de recomendación juegan un papel crítico en muchos negocios, especialmente en los de comercio electrónico y medios de comunicación, ya que impactan directamente en los ingresos por ventas y en el compromiso de los usuarios.
Generalmente, hay dos maneras de desarrollar una lista de recomendaciones:

1. Filtrado colaborativo
El método de filtrado colaborativo se basa en los comportamientos anteriores de los usuarios, como las páginas que han visto, los productos que han comprado o las valoraciones que han dado a diferentes artículos. El método de filtrado colaborativo utiliza estos datos para encontrar similitudes entre usuarios o artículos, y recomienda a los usuarios los artículos o contenidos más similares.

La hipótesis básica del método de filtrado colaborativo es que quienes han visto o comprado contenidos o productos similares en el pasado probablemente verán o comprarán tipos de contenidos o productos similares en el futuro.

Así, basándose en esta suposición, si una persona compró los artículos A, B y C y otra persona compró los artículos A, B y D en el pasado, es probable que la primera persona compre el artículo D y la otra persona compre el artículo C, ya que comparten muchas similitudes entre sí

2. Filtrado basado en el contenido
Por otro lado, el filtrado basado en el contenido produce una lista de recomendaciones basada en las características de un artículo o de un usuario. Suele examinar las palabras clave que describen las características de un artículo. La hipótesis básica del método de filtrado basado en el contenido es que los usuarios probablemente verán o comprarán artículos similares a los que han comprado o visto en el pasado

Por ejemplo, si un usuario ha escuchado algunas canciones en el pasado, el método de filtrado basado en el contenido recomendará tipos de canciones similares que comparten características parecidas a las que el usuario ya ha escuchado.

Construcción de un sistema de recomendación de productos con filtrado colaborativo
Como se ha mencionado, un algoritmo de filtrado colaborativo se utiliza para recomendar productos basándose en el historial de comportamiento de los usuarios y en las similitudes entre ellos. El primer paso para implementar un algoritmo de filtrado colaborativo para un sistema de recomendación de productos es construir una matriz usuario-elemento

Una matriz usuario-elemento comprende usuarios individuales en las filas y elementos individuales en las columnas. Será más fácil de explicar con un ejemplo. Observe la siguiente matriz.


Las filas de esta matriz representan a cada usuario y las columnas a cada elemento. Los valores de cada celda representan si el usuario en cuestión ha comprado o no el elemento en cuestión. Por ejemplo, el usuario 1 ha comprado los artículos B y D y el usuario 2 ha comprado los artículos A, B, C y E

Para construir un sistema de recomendación de productos basado en el filtrado colaborativo, necesitamos construir primero este tipo de matriz usuario-elemento. Con esta matriz usuario-elemento, el siguiente paso para construir un sistema de recomendación de productos basado en el filtrado colaborativo es calcular las similitudes entre los usuarios

Para medir las similitudes, se suele utilizar la similitud de los cosenos. La ecuación para calcular la similitud del coseno entre dos usuarios es la siguiente

En esta ecuación, U1 y U2 representan al usuario 1 y al usuario 2. P1i y P2i representan cada producto, i, que el usuario 1 y el usuario 2 han comprado. Si utilizas esta ecuación, obtendrás 0,353553 como la similitud del coseno entre los usuarios 1 y 2 en el ejemplo anterior y 0,866025 como la similitud del coseno entre los usuarios 2 y 4.

Como puede imaginar, cuanto mayor sea la similitud del coseno, más parecidos serán los dos usuarios. Así, en nuestro ejemplo, los usuarios 2 y 4 son más similares entre sí que los usuarios 1 y 2. Por último, cuando se utiliza un algoritmo de filtrado colaborativo para las recomendaciones de productos, se pueden adoptar dos enfoques: un enfoque basado en el usuario y un enfoque basado en el elemento

Como sus nombres sugieren, el enfoque del filtrado colaborativo basado en el usuario utiliza las similitudes entre los usuarios. Por otro lado, el enfoque de filtrado colaborativo basado en el elemento utiliza las similitudes entre los elemento. Esto significa que cuando calculamos las similitudes entre los dos usuarios en el filtrado colaborativo del enfoque basado en el usuario, tenemos que construir y utilizar una matriz usuario-elemento, como hemos comentado anteriormente

Sin embargo, para el enfoque basado en elementos, necesitamos calcular las similitudes entre los dos elementos, y esto significa que necesitamos construir y utilizar una matriz de elemento a usuario, que podemos obtener simplemente transponiendo la matriz de usuario a elemento.

In [1]:
import pandas as pd

UTILIZAR ESTA PARTE SI ESTAS EN GOOGLE COLAB

In [ ]:
import sys
assert sys.version_info >= (3, 5)
import os

In [ ]:
RETAIL_PATH = "https://github.com/hcgalvan/UNSAM-Machine-Learning-on-Economics/raw/main/data/"

In [ ]:
if 'google.colab' in sys.modules:
  def load_datasets_h1(datasets_path=RETAIL_PATH):
    xlsx_path = os.path.join(datasets_path, "online_retail_II.xlsx")
    return pd.read_excel(xlsx_path)

  # The function loads the data as a Pandas DataFrame instance.
  retail_ol_h1 = load_datasets_h1()
  def load_datasets_h2(datasets_path=RETAIL_PATH):
      xlsx_path = os.path.join(datasets_path, "online_retail_II.xlsx")
      return pd.read_excel(xlsx_path,1)

  # The function loads the data as a Pandas DataFrame instance.
  retail_ol_h2 = load_datasets_h2()

# 1. Load Data

UTILIZAR ESTE PARTE DEL CODIGO SOLO SI DESCARGASTE EN PC

In [3]:
retail_ol_h1= pd.read_excel(io='./data/online_retail_II.xlsx', sheet_name='Year 2010-2011')
retail_ol_h2= pd.read_excel(io='./data/online_retail_II.xlsx', sheet_name='Year 2009-2010')

LO QUE SIGUE ES PARTE COMUN.
Para trabajar en googlecolab como para los que hemos descargado en PC.

In [4]:
frames = [retail_ol_h1, retail_ol_h2]
results = pd.concat(frames)
df = results.copy()

In [5]:
df.shape

(1067371, 8)

In [6]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  
0 2010-12-01 08:26:00   2.55      17850.0  United Kingdom  
1 2010-12-01 08:26:00   3.39      17850.0  United Kingdom  
2 2010-12-01 08:26:00   2.75      17850.0  United Kingdom  
3 2010-12-01 08:26:00   3.39      17850.0  United Kingdom  
4 2010-12-01 08:26:00   3.39      17850.0  United Kingdom


Hay registros con valores negativos en la columna Cantidad, que representan pedidos cancelados. Ignoremos y eliminemos estos registros. Podemos filtrar todos estos registros en nuestro DataFrame simplemente eligiendo los Quantity que sean mayores a 0

In [7]:
df = df.loc[df['Quantity'] > 0]

# 2. Data Preparation

Preparación de los datos
Antes de sumergirnos en la construcción de un motor de recomendación de productos utilizando un algoritmo de filtrado colaborativo, tenemos que hacer el siguiente par de cosas:

Gestionar los valores NaN en el conjunto de datos
Construir una matriz usuario-elemento
En primer lugar, tenemos que gestionar los valores "NaN" de nuestro conjunto de datos, especialmente los "NaN" del campo "Customer ID". Sin los valores correctos en el campo Customer ID, no podemos construir un sistema de recomendación adecuado, ya que el algoritmo de filtrado colaborativo depende de los datos históricos de compra de artículos de clientes individuales.

En segundo lugar, necesitamos construir una matriz usuario-elemento antes de poder implementar el algoritmo de filtrado colaborativo para la recomendación de productos. La matriz usuario-elemento es simplemente una tabla de datos, en la que cada columna representa cada producto o elemento, cada fila representa un cliente y el valor de cada celda representa si el cliente en cuestión ha comprado o no el producto en cuestión.

Gestión de los NaN en el campo Customer ID

Si observamos los datos con detenimiento, nos daremos cuenta de que hay algunos registros sin identificación de cliente. Como tenemos que construir una matriz de elementos de clientes en la que cada fila es específica para cada cliente, no podemos incluir esos registros sin CustomerID en nuestros datos.

#### - Handle NaNs in CustomerID field

In [8]:
df['Customer ID'].describe()

count    805620.000000
mean      15331.856250
std        1696.768395
min       12346.000000
25%       13982.000000
50%       15271.000000
75%       16805.000000
max       18287.000000
Name: Customer ID, dtype: float64

Veamos primero cuántos registros no tienen identificación de cliente. Haga una sumatoria de todos los valores nan para CustomerID en la siguiente celda

In [9]:
df['Customer ID'].isna().sum()

238801

Como puede ver en esta salida, hay 238.801 registros sin identificación de cliente. Y algunos de los datos a los que les falta la identificación del cliente tienen este aspecto:

In [10]:
df.loc[df['Customer ID'].isna()].head()

Invoice StockCode                      Description  Quantity  \
622   536414     22139                              NaN        56   
1443  536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444  536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445  536544     21786               POLKADOT RAIN HAT          4   
1446  536544     21787            RAIN PONCHO RETROSPOT         2   

             InvoiceDate  Price  Customer ID         Country  
622  2010-12-01 11:52:00   0.00          NaN  United Kingdom  
1443 2010-12-01 14:32:00   2.51          NaN  United Kingdom  
1444 2010-12-01 14:32:00   2.51          NaN  United Kingdom  
1445 2010-12-01 14:32:00   0.85          NaN  United Kingdom  
1446 2010-12-01 14:32:00   1.66          NaN  United Kingdom

In [11]:
df.shape

(1044421, 8)

Ahora que sabemos que hay registros con entradas de identificación de clientes que faltan, tenemos que excluirlos del análisis posterior. Una forma de eliminarlos de nuestro DataFrame es utilizar la función dropna. Utiliza dicho metodo para eliminarlos
Auito los NaN utilizando de filtro a la columna Customer ID 

In [12]:
df = df.dropna(subset=['Customer ID'])

In [13]:
df.shape

(805620, 8)

In [14]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  
0 2010-12-01 08:26:00   2.55      17850.0  United Kingdom  
1 2010-12-01 08:26:00   3.39      17850.0  United Kingdom  
2 2010-12-01 08:26:00   2.75      17850.0  United Kingdom  
3 2010-12-01 08:26:00   3.39      17850.0  United Kingdom  
4 2010-12-01 08:26:00   3.39      17850.0  United Kingdom

#### - Customer-Item Matrix

Construir una matriz de elementos para el cliente
Los datos que tenemos ahora representan artículos individuales comprados por los clientes. Sin embargo, para construir un sistema de recomendación de productos con un algoritmo de filtrado colaborativo, necesitamos tener datos en los que cada registro contenga información sobre qué artículo ha comprado cada cliente.

Vamos a transformar los datos en una matriz usuario-elemento, donde cada fila representa un cliente y las columnas corresponden a diferentes productos.

En la siguiente celda, hacemos un pivot_table de pandas con los siguientes parametros:

In [15]:
customer_item_matrix = df.pivot_table(
    index='Customer ID', 
    columns='StockCode', 
    values='Quantity',
    aggfunc='sum'
)

Y asignalo a la variable customer_item_matrix

In [16]:
customer_item_matrix.loc[12481:].head()

StockCode    10002  10080  10109  10120  10125  10133  10134  10135  10138  \
Customer ID                                                                  
12481.0        NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
12482.0        NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
12483.0        NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
12484.0        NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
12485.0        NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   

StockCode    11001  ...  BANK CHARGES  C2   D  DOT    M  PADS  POST  SP1002  \
Customer ID         ...                                                       
12481.0        NaN  ...           NaN NaN NaN  NaN  NaN   NaN  49.0     NaN   
12482.0        NaN  ...           NaN NaN NaN  NaN  2.0   NaN   NaN     NaN   
12483.0        NaN  ...           NaN NaN NaN  NaN  NaN   NaN  29.0     NaN   
12484.0       16.0  ...           NaN NaN NaN  NaN  NaN   NaN  29.0     NaN   
12485.0        NaN  ...           NaN NaN NaN  NaN  NaN   NaN   NaN     NaN   

StockCode    TEST001  TEST002  
Customer ID                    
12481.0          NaN      NaN  
12482.0          NaN      NaN  
12483.0          NaN      NaN  
12484.0          NaN      NaN  
12485.0          NaN      NaN  

[5 rows x 4631 columns]


Como podemos ver, ahora tenemos una matriz en la que cada fila representa las cantidades totales compradas de cada producto para cada cliente.

Ahora, codifiquemos 0-1 estos datos, de manera que un valor de 1 significa que el producto dado fue comprado por el cliente dado, y un valor de 0 significa que el producto dado nunca fue comprado por el cliente dado. 
Este procedimiento puede realizarse de varias formas, le recomendamos hacerlo con una función lambda y con el metodo applymap de pandas. Asigne el resultado a la misma variable anterior customer_item_matrix

Estamos recomendando la función "applymap", que aplica una función determinada a cada elemento de un DataFrame.

La función Lambda recomendanda en este código simplemente codifica todos los elementos cuyos valores son mayores que 0 con 1, y el resto con 0.

In [17]:
customer_item_matrix.shape

(5881, 4631)

In [18]:
df['StockCode'].nunique()

4631

In [19]:
df['Customer ID'].nunique()

5881

In [20]:
customer_item_matrix.loc[12348.0].sum()

2714.0

In [21]:
customer_item_matrix = customer_item_matrix.applymap(lambda x: 1 if x > 0 else 0)

In [22]:
customer_item_matrix.loc[12481:].head()

StockCode    10002  10080  10109  10120  10125  10133  10134  10135  10138  \
Customer ID                                                                  
12481.0          0      0      0      0      0      0      0      0      0   
12482.0          0      0      0      0      0      0      0      0      0   
12483.0          0      0      0      0      0      0      0      0      0   
12484.0          0      0      0      0      0      0      0      0      0   
12485.0          0      0      0      0      0      0      0      0      0   

StockCode    11001  ...  BANK CHARGES  C2  D  DOT  M  PADS  POST  SP1002  \
Customer ID         ...                                                    
12481.0          0  ...             0   0  0    0  0     0     1       0   
12482.0          0  ...             0   0  0    0  1     0     0       0   
12483.0          0  ...             0   0  0    0  0     0     1       0   
12484.0          1  ...             0   0  0    0  0     0     1       0   
12485.0          0  ...             0   0  0    0  0     0     0       0   

StockCode    TEST001  TEST002  
Customer ID                    
12481.0            0        0  
12482.0            0        0  
12483.0            0        0  
12484.0            0        0  
12485.0            0        0  

[5 rows x 4631 columns]

Ahora tenemos un array de elementos del cliente que podemos utilizar para el algoritmo de filtrado colaborativo. Pasemos ahora a la construcción de motores de recomendación de productos.

Filtrado colaborativo
Exploraremos dos enfoques para construir nuestro recomendador basado en el usuario y en el elemento. En el enfoque basado en el usuario, calculamos las similitudes entre los usuarios basándonos en su historial de compras de artículos. En el enfoque basado en los artículos, por otro lado, calculamos las similitudes entre los artículos en función de qué artículos se compran a menudo junto con otros.

Para medir la similitud entre usuarios o entre artículos, utilizaremos el método cosine_similarity del paquete scikit-learn. 
Importemos esta función utilizando el siguiente código:

# 3. Collaborative Filtering

Esta función cosine_similarity del paquete sklearn calcula las similitudes del coseno en pares en los datos dados.

In [24]:
from sklearn.metrics.pairwise import cosine_similarity

## 3.1. User-based Collaborative Filtering

Para construir un algoritmo de filtrado colaborativo basado en el usuario, necesitamos calcular las similitudes del coseno entre los usuarios. Para ello hacemos lo siguiente: asignar a una variable llamada user_user_sim_matrix un nuevo dataframe de pandas, que podemos crear con el comando pd.DataFrame() y pasarle como parametro la funcion de scikit-learn que acabamos de importar y pasandole como parametro nuestra matriz customer_item_matrix

#### - User-to-User Similarity Matrix

In [30]:
user_user_sim_matrix = pd.DataFrame(
    cosine_similarity(customer_item_matrix)
)

In [31]:
user_user_sim_matrix.head()

0         1         2         3         4         5         6     \
0  1.00000  0.000000  0.000000  0.131060  0.000000  0.000000  0.023002   
1  0.00000  1.000000  0.053452  0.045502  0.043214  0.038881  0.031944   
2  0.00000  0.053452  1.000000  0.017025  0.048507  0.000000  0.023905   
3  0.13106  0.045502  0.017025  1.000000  0.041292  0.037152  0.152617   
4  0.00000  0.043214  0.048507  0.041292  1.000000  0.000000  0.028989   

       7         8         9     ...      5871  5872      5873      5874  \
0  0.000000  0.000000  0.000000  ...  0.000000   0.0  0.000000  0.000000   
1  0.055728  0.023395  0.090351  ...  0.037987   0.0  0.067344  0.064820   
2  0.000000  0.026261  0.067612  ...  0.000000   0.0  0.000000  0.000000   
3  0.071000  0.044710  0.057555  ...  0.054447   0.0  0.032174  0.020646   
4  0.050572  0.000000  0.000000  ...  0.000000   0.0  0.000000  0.058824   

       5875      5876      5877  5878      5879      5880  
0  0.000000  0.000000  0.000000   0.0  0.070535  0.000000  
1  0.102869  0.113961  0.067344   0.0  0.076186  0.024398  
2  0.000000  0.138580  0.000000   0.0  0.000000  0.054772  
3  0.049147  0.140654  0.016087   0.0  0.062399  0.038854  
4  0.000000  0.038782  0.000000   0.0  0.029630  0.000000  

[5 rows x 5881 columns]

Como podemos ver en este código, estamos utilizando la función cosine_similarity del módulo metrics.pairwise del paquete sklearn. Esta función calcula las similitudes del coseno en pares entre las muestras y produce los resultados como un tipo de array.

A continuación, creamos un pandasDataFrame con este array de salida y lo almacenamos en una variable llamada user_user_sim_matrix, que significa array de similitudes usuario-usuario.

Como puedes ver, el índice y los nombres de las columnas no son fáciles de entender. Como cada columna y fila del índice representa a clientes individuales, vamos a renombrar el índice y las columnas usando el siguiente código:

In [32]:
user_user_sim_matrix.columns = customer_item_matrix.index

user_user_sim_matrix['Customer ID'] = customer_item_matrix.index
user_user_sim_matrix = user_user_sim_matrix.set_index('Customer ID')

In [33]:
user_user_sim_matrix.head()

Customer ID  12346.0   12347.0   12348.0   12349.0   12350.0   12351.0  \
Customer ID                                                              
12346.0      1.00000  0.000000  0.000000  0.131060  0.000000  0.000000   
12347.0      0.00000  1.000000  0.053452  0.045502  0.043214  0.038881   
12348.0      0.00000  0.053452  1.000000  0.017025  0.048507  0.000000   
12349.0      0.13106  0.045502  0.017025  1.000000  0.041292  0.037152   
12350.0      0.00000  0.043214  0.048507  0.041292  1.000000  0.000000   

Customer ID   12352.0   12353.0   12354.0   12355.0  ...   18278.0  18279.0  \
Customer ID                                          ...                      
12346.0      0.023002  0.000000  0.000000  0.000000  ...  0.000000      0.0   
12347.0      0.031944  0.055728  0.023395  0.090351  ...  0.037987      0.0   
12348.0      0.023905  0.000000  0.026261  0.067612  ...  0.000000      0.0   
12349.0      0.152617  0.071000  0.044710  0.057555  ...  0.054447      0.0   
12350.0      0.028989  0.050572  0.000000  0.000000  ...  0.000000      0.0   

Customer ID   18280.0   18281.0   18282.0   18283.0   18284.0  18285.0  \
Customer ID                                                              
12346.0      0.000000  0.000000  0.000000  0.000000  0.000000      0.0   
12347.0      0.067344  0.064820  0.102869  0.113961  0.067344      0.0   
12348.0      0.000000  0.000000  0.000000  0.138580  0.000000      0.0   
12349.0      0.032174  0.020646  0.049147  0.140654  0.016087      0.0   
12350.0      0.000000  0.058824  0.000000  0.038782  0.000000      0.0   

Customer ID   18286.0   18287.0  
Customer ID                      
12346.0      0.070535  0.000000  
12347.0      0.076186  0.024398  
12348.0      0.000000  0.054772  
12349.0      0.062399  0.038854  
12350.0      0.029630  0.000000  

[5 rows x 5881 columns]

Veamos con más detalle esta matriz de similitudes entre usuarios. Como se puede imaginar, la similitud del coseno entre un cliente y él mismo es 1, y esto es lo que podemos observar en esta matriz de similitud. Los elementos diagonales de esta matriz de similitudes entre usuarios tienen valores de 1.

El resto representa la similitud del coseno entre dos clientes. Por ejemplo, la medida de similitud del coseno entre los clientes 12347 y 12348 es 0,063022. Por otro lado, la similitud del coseno entre los clientes "12347" y "12349" es de 0,046130. Esto sugiere que el cliente 12348 es más similar al cliente 12347 que el cliente 12349, basándose en los productos que compraron. De este modo, podemos saber fácilmente qué clientes son similares a otros, y qué clientes han comprado artículos similares a otros.

Estas medidas de similitud de pares de coseno son las que utilizaremos para las recomendaciones de productos. Vamos a trabajar eligiendo un cliente como ejemplo. Primero clasificaremos los clientes más parecidos al cliente con ID 12350, para hacer esto

con iloc selecciona el id con valor 12350.0
una vez seleccionado organizalos en orden descendente con el metodo sort_values de pandas

#### - Making Recommendations

In [35]:
user_user_sim_matrix.loc[12350.0].sort_values(ascending=False)

Customer ID
12350.0    1.000000
12636.0    0.242536
12568.0    0.216930
12503.0    0.171499
16886.0    0.171499
             ...   
14960.0    0.000000
14962.0    0.000000
14964.0    0.000000
14967.0    0.000000
18287.0    0.000000
Name: 12350.0, Length: 5881, dtype: float64

In [37]:
items_bought_by_A = set(customer_item_matrix.loc[12350.0].iloc[
    customer_item_matrix.loc[12350.0].to_numpy().nonzero()
].index)

In [38]:
items_bought_by_A = set(customer_item_matrix.loc[12350.0].iloc[
    customer_item_matrix.loc[12350.0].to_numpy().nonzero()
].index)
items_bought_by_A

{20615,
 20652,
 21171,
 21832,
 21864,
 21866,
 21908,
 21915,
 22348,
 22412,
 22551,
 22557,
 22620,
 '79066K',
 '79191C',
 '84086C',
 'POST'}

In [39]:
items_bought_by_B = set(customer_item_matrix.loc[17935.0].iloc[
    customer_item_matrix.loc[17935.0].to_numpy().nonzero()
].index)
items_bought_by_B

{20657,
 20659,
 20713,
 20828,
 20856,
 20983,
 21051,
 21866,
 21867,
 21930,
 21931,
 21934,
 22200,
 22202,
 22208,
 22209,
 22210,
 22211,
 22304,
 22305,
 22423,
 22449,
 22450,
 22473,
 22474,
 22475,
 22513,
 22551,
 22553,
 22557,
 22633,
 22640,
 22654,
 22659,
 22697,
 22698,
 22699,
 22749,
 22752,
 22753,
 22754,
 22755,
 22759,
 22834,
 22837,
 22890,
 22906,
 23290,
 23292,
 23309,
 '47599B',
 '48173C',
 '75149N',
 '82001S',
 '85049E',
 '85099B',
 '85099C',
 85214,
 '85232B',
 'POST'}

Ahora tenemos dos conjuntos de artículos que los clientes 12350 y 17935 han comprado. Utilizando una simple operación de conjunto, podemos encontrar los artículos que el cliente 12350 ha comprado, pero el cliente 17935 no. Haz una resta entre items_bought_by_A y items_bought_by_B y asignalo a la variable items_to_recommend_to_B

In [40]:
items_to_recommend_to_B = items_bought_by_A - items_bought_by_B

In [41]:
items_to_recommend_to_B

{20615,
 20652,
 21171,
 21832,
 21864,
 21908,
 21915,
 22348,
 22412,
 22620,
 '79066K',
 '79191C',
 '84086C'}

Para obtener las descripciones de estos elementos, podemos utilizar el siguiente código:

In [42]:
df.loc[
    df['StockCode'].isin(items_to_recommend_to_B), 
    ['StockCode', 'Description']
].drop_duplicates().set_index('StockCode')

Description
StockCode                                 
21832                 CHOCOLATE CALCULATOR
21915               RED  HARMONICA IN BOX 
22620          4 TRADITIONAL SPINNING TOPS
79066K                      RETRO MOD TRAY
21864      UNION JACK FLAG PASSPORT COVER 
79191C         RETRO PLASTIC ELEPHANT TRAY
21908        CHOCOLATE THIS WAY METAL SIGN
20615         BLUE POLKADOT PASSPORT COVER
20652           BLUE POLKADOT LUGGAGE TAG 
22348          TEA BAG PLATE RED RETROSPOT
22412      METAL SIGN NEIGHBOURHOOD WITCH 
21171                 BATHROOM METAL SIGN 
84086C             PINK/PURPLE RETRO RADIO
22348            TEA BAG PLATE RED SPOTTY 
20652             BLUE SPOTTY LUGGAGE TAG 
20615           BLUE SPOTTY PASSPORT COVER

## 3.2. Item-based Collaborative Filtering

Con este mismo ejercicio podemos hacer un filtrado colaborativo basado en elementos o items.

#### - Item-to-Item Similarity Matrix

In [43]:
item_item_sim_matrix = pd.DataFrame(cosine_similarity(customer_item_matrix.T))

In [44]:
item_item_sim_matrix.columns = customer_item_matrix.T.index

item_item_sim_matrix['StockCode'] = customer_item_matrix.T.index
item_item_sim_matrix = item_item_sim_matrix.set_index('StockCode')

In [45]:
item_item_sim_matrix

StockCode     10002     10080  10109     10120     10125     10133     10134  \
StockCode                                                                      
10002      1.000000  0.032564    0.0  0.075801  0.100462  0.063130  0.035316   
10080      0.032564  1.000000    0.0  0.028916  0.067065  0.033715  0.031435   
10109      0.000000  0.000000    1.0  0.000000  0.000000  0.080845  0.000000   
10120      0.075801  0.028916    0.0  1.000000  0.059470  0.044845  0.000000   
10125      0.100462  0.067065    0.0  0.059470  1.000000  0.060673  0.032325   
...             ...       ...    ...       ...       ...       ...       ...   
PADS       0.000000  0.000000    0.0  0.000000  0.053606  0.020211  0.000000   
POST       0.085364  0.000000    0.0  0.013782  0.058601  0.020086  0.000000   
SP1002     0.110432  0.000000    0.0  0.098058  0.075810  0.057166  0.000000   
TEST001    0.000000  0.000000    0.0  0.000000  0.000000  0.000000  0.000000   
TEST002    0.000000  0.000000    0.0  0.000000  0.000000  0.000000  0.000000   

StockCode     10135     10138     11001  ...  BANK CHARGES        C2  \
StockCode                                ...                           
10002      0.076087  0.062609  0.071885  ...      0.035829  0.011641   
10080      0.046886  0.000000  0.019195  ...      0.000000  0.000000   
10109      0.074953  0.267261  0.000000  ...      0.000000  0.000000   
10120      0.051971  0.037062  0.089363  ...      0.000000  0.000000   
10125      0.072322  0.000000  0.078957  ...      0.024596  0.015982   
...             ...       ...       ...  ...           ...       ...   
PADS       0.018738  0.000000  0.023014  ...      0.000000  0.000000   
POST       0.029796  0.000000  0.059467  ...      0.022800  0.014815   
SP1002     0.053000  0.188982  0.065094  ...      0.000000  0.000000   
TEST001    0.000000  0.000000  0.000000  ...      0.000000  0.000000   
TEST002    0.000000  0.000000  0.000000  ...      0.000000  0.000000   

StockCode         D  DOT         M      PADS      POST    SP1002  TEST001  \
StockCode                                                                   
10002      0.000000  0.0  0.144691  0.000000  0.085364  0.110432      0.0   
10080      0.000000  0.0  0.059441  0.000000  0.000000  0.000000      0.0   
10109      0.000000  0.0  0.047511  0.000000  0.000000  0.000000      0.0   
10120      0.000000  0.0  0.112007  0.000000  0.013782  0.098058      0.0   
10125      0.000000  0.0  0.101875  0.053606  0.058601  0.075810      0.0   
...             ...  ...       ...       ...       ...       ...      ...   
PADS       0.000000  0.0  0.000000  1.000000  0.000000  0.000000      0.0   
POST       0.022222  0.0  0.082630  0.000000  1.000000  0.035136      0.0   
SP1002     0.000000  0.0  0.067191  0.000000  0.035136  1.000000      0.0   
TEST001    0.000000  0.0  0.023756  0.000000  0.000000  0.000000      1.0   
TEST002    0.000000  0.0  0.000000  0.000000  0.000000  0.000000      0.5   

StockCode  TEST002  
StockCode           
10002          0.0  
10080          0.0  
10109          0.0  
10120          0.0  
10125          0.0  
...            ...  
PADS           0.0  
POST           0.0  
SP1002         0.0  
TEST001        0.5  
TEST002        1.0  

[4631 rows x 4631 columns]

#### - Making Recommendations

In [46]:
top_10_similar_items = list(
    item_item_sim_matrix\
        .loc[23166]\
        .sort_values(ascending=False)\
        .iloc[:10]\
    .index
)

In [47]:
top_10_similar_items

[23166, 23165, 23167, 22993, 23307, 22720, 22722, 23243, 23306, 22961]

In [48]:
df.loc[
    df['StockCode'].isin(top_10_similar_items), 
    ['StockCode', 'Description']
].drop_duplicates().set_index('StockCode').loc[top_10_similar_items]

Description
StockCode                                     
23166           MEDIUM CERAMIC TOP STORAGE JAR
23165            LARGE CERAMIC TOP STORAGE JAR
23167           SMALL CERAMIC TOP STORAGE JAR 
22993             SET OF 4 PANTRY JELLY MOULDS
23307      SET OF 60 PANTRY DESIGN CAKE CASES 
22720        SET OF 3 CAKE TINS PANTRY DESIGN 
22722        SET OF 6 SPICE TINS PANTRY DESIGN
23243      SET OF TEA COFFEE SUGAR TINS PANTRY
23306          SET OF 36 DOILIES PANTRY DESIGN
23306           SET OF 36 PANTRY PAPER DOILIES
22961                   JAM MAKING SET PRINTED

De esta forma construimos los dos sistemas de recomendaciones.